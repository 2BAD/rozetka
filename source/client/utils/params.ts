import { OptionsOfTextResponseBody as Params } from 'got'

/**
 * Create Options object with searchParams populated
 */
export const buildParams = (params: string[][]): Params => ({
  searchParams: new URLSearchParams(params)
})
