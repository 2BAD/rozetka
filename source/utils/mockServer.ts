import { URL } from 'url'
import nock, { Definition } from 'nock'

type Mock = {
  done: () => void
  context: nock.BackContext
}

/**
 * Mock server with fixture recording and playback support
 * @param url context specific module url
 * @param file file name for stub object
 */
export const server = (url: string, file: string): Promise<Mock> => {
  nock.back.fixtures = new URL('./__fixtures__/', url).pathname

  return nock
    .back(file, {
      before,
      afterRecord
    })
    .then(({ nockDone, context }) => ({
      done: () => {
        // restore the HTTP interceptor to the normal unmocked behaviour
        nock.restore()
        nockDone()
      },
      context
    }))
}

/**
 * Match end replace any occurrences of `r=*` in query string
 */
const stripTimestamp = (path: string): string => {
  return path.replace(/(?<=\?|&)r=[^&]*/g, '')
}

const before = (scope: Definition): void => {
  // @ts-expect-error it does exist
  scope.filteringPath = stripTimestamp
}

const afterRecord = (fixtures: Definition[]): Definition[] =>
  fixtures.map((f) => {
    f.path = stripTimestamp(f.path)
    return f
  })
