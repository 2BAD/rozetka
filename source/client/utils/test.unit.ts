import { params } from './params'

describe('parameter builder', () => {
  it('should return Options object with searchParam set to URLSearchParams object instance', () => {
    expect.assertions(3)

    const p = params(['goodsId', '100'])

    expect(p).toHaveProperty('searchParams')
    expect(p.searchParams).toBeInstanceOf(URLSearchParams)
    expect(p).toMatchSnapshot()
  })

  it('should accept an array of arrays', () => {
    expect.assertions(2)

    const p = params(['goodsId', '100'], ['lang', 'ru'])

    expect(p.searchParams?.toString()).toBe('goodsId=100&lang=ru')
    expect(p).toMatchSnapshot()
  })
})
