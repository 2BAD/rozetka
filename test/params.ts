import { buildParams } from '../source/client/utils/params'

describe('Parameter builder for Got client', () => {
  it('should return Options object with searchParam set to URLSearchParams object instance', () => {
    expect.assertions(3)

    const params = buildParams([['goodsId', '100']])

    expect(params).toHaveProperty('searchParams')
    expect(params.searchParams).toBeInstanceOf(URLSearchParams)
    expect(params).toMatchSnapshot()
  })

  it('should accept array of arrays', () => {
    expect.assertions(2)

    const params = buildParams([
      ['goodsId', '100'],
      ['lang', 'ru']
    ])

    expect(params.searchParams?.toString()).toBe('goodsId=100&lang=ru')
    expect(params).toMatchSnapshot()
  })
})
