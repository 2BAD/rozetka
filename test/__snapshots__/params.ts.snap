// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parameter builder for Got client should accept array of arrays 1`] = `
Object {
  "searchParams": URLSearchParams {
    Symbol(query): Array [
      "goodsId",
      "100",
      "lang",
      "ru",
    ],
    Symbol(context): null,
  },
}
`;

exports[`Parameter builder for Got client should return Options object with searchParam set to URLSearchParams object instance 1`] = `
Object {
  "searchParams": URLSearchParams {
    Symbol(query): Array [
      "goodsId",
      "100",
    ],
    Symbol(context): null,
  },
}
`;
